### PR TITLE
Four tag and perspective transformation alongside affine transformation of three tag sources.

### PR DESCRIPTION
**Pull request by JohanBlome**

### CODE REVIEW

File: src/aruco_common.py [hunk 1]
```diff
@@ -57,13 +59,13 @@ def detect_aruco_tags(img, aruco_dict_id=ARUCO_DICT_ID):
     # convert input to grayscale
     gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
     # locate the aruco markers
-    try:
+    if CV_VERSION[0] <= 4 and CV_VERSION[1] < 7:
```
Comment (chemag) [inline src/aruco_common.py:62]: Nice, much better than before

File: src/vft.py [hunk 2]
```diff
@@ -91,9 +91,18 @@ def generate_graycode(width, height, vft_id, tag_border_size, value, debug):
...
+    )
     # convert gray code in bit_stream to a number
-    num_read = gray_bitstream_to_num(bit_stream)
+    try:
+        num_read = gray_bitstream_to_num(bit_stream)
+    except (InvalidGrayCode, SingleGraycodeBitError, NoValidTag) as e:
+        print(f"{type(e).__name__}: {e}")
+        if lock_layout and vft_id:
+            # Rerun without locking the layout
+            return analyze_graycode(img, luma_threshold, lock_layout=False, debug=debug)
```
Comment (chemag) [inline src/vft.py:106]: Won't this run infinitely?

File: src/vft.py [hunk 3]
```diff
@@ -134,8 +142,9 @@ def generate(width, height, vft_id, tag_border_size, value, debug):
     vft_layout = VFTLayout(width, height, vft_id, tag_border_size)
     # 2. add fiduciary markers (tags) in the top-left, top-right,
     # and bottom-left corners
-    for tag_number in range(3):
+    for tag_number in range(4):
```
Comment (chemag) [inline src/vft.py]: can you please move this 4 to a constant?

File: src/metiq.py [hunk 4]
```diff
@@ -1179,6 +1179,7 @@ def main(argv):
                     cache_audio=cache_audio,
```
Comment (chemag) [inline src/metiq.py:1179]: suggest renaming the patch to:

metiq: upgrade affine transformations to homographies

I understood immediately the "perspective transformation" term, but my reading suggests "homography" is a more common term in English

**Review by chemag — COMMENTED** [review]
nits

File: src/vft.py (hunk 2 above)
Comment (JohanBlome) [reply src/vft.py:106]: No. It only runs if lock_layout is enabled which in this single recursion it is not.
The idea here is that when the parsing fails it tries again without the locked layout setting.

File: src/vft.py (hunk 3 above)
Comment (JohanBlome) [reply src/vft.py]: Good idea.

File: src/metiq.py (hunk 4 above)
Comment (JohanBlome) [reply src/metiq.py:1179]: Haha, at least I learned a new word as well. OK. I'll do.

**Pull request merged**